keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/CombinedRepository.kt

```diff
@@ -0,0 +1,387 @@
+package com.netflix.spinnaker.keel.persistence
+
+import com.netflix.spinnaker.keel.api.DeliveryConfig
+import com.netflix.spinnaker.keel.api.Environment
+import com.netflix.spinnaker.keel.api.Resource
+import com.netflix.spinnaker.keel.api.ResourceSpec
+import com.netflix.spinnaker.keel.api.artifacts.ArtifactStatus
+import com.netflix.spinnaker.keel.api.artifacts.ArtifactType
+import com.netflix.spinnaker.keel.api.artifacts.DebianArtifact
+import com.netflix.spinnaker.keel.api.artifacts.DeliveryArtifact
+import com.netflix.spinnaker.keel.api.artifacts.DockerArtifact
+import com.netflix.spinnaker.keel.api.id
+import com.netflix.spinnaker.keel.constraints.ConstraintState
+import com.netflix.spinnaker.keel.core.api.EnvironmentArtifactPin
+import com.netflix.spinnaker.keel.core.api.EnvironmentArtifactVetoes
+import com.netflix.spinnaker.keel.core.api.EnvironmentArtifactsSummary
+import com.netflix.spinnaker.keel.core.api.PinnedEnvironment
+import com.netflix.spinnaker.keel.core.api.SubmittedDeliveryConfig
+import com.netflix.spinnaker.keel.core.api.UID
+import com.netflix.spinnaker.keel.core.api.normalize
+import com.netflix.spinnaker.keel.core.api.resources
+import com.netflix.spinnaker.keel.diff.DefaultResourceDiff
+import com.netflix.spinnaker.keel.events.ArtifactRegisteredEvent
+import com.netflix.spinnaker.keel.events.ResourceCreated
+import com.netflix.spinnaker.keel.events.ResourceEvent
+import com.netflix.spinnaker.keel.events.ResourceUpdated
+import com.netflix.spinnaker.keel.exceptions.DuplicateResourceIdException
+import java.time.Clock
+import java.time.Duration
+import org.slf4j.LoggerFactory
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.stereotype.Component
+import org.springframework.transaction.annotation.Propagation
+import org.springframework.transaction.annotation.Transactional
+
+/**
+ * A combined repository for delivery configs, artifacts, and resources.
+ *
+ * This paves the way for re-thinking how we interact with sql/storing of resources,
+ * and updating our internal repository structure to allow storing delivery configs to more easily also
+ * store artifacts and resources.
+ *
+ * This also gives us an easy place to emit telemetry and events around the usage of methods.
+ *
+ * TODO eb: refactor repository interaction so transactionality is easier.
+ */
+@Component
+class CombinedRepository(
```

Comment:
Consider making a `KeelRepository` interface covering all of the public methods here, and use that type to inject `CombinedRepository` everywhere needed.

Reply:
Done!